### PR TITLE
address an issue when using ArgParser.allowAnything() with CommandRunner

### PR DESCRIPTION
**Pull request by devoncarew**
Address an issue when using `ArgParser.allowAnything()` with `CommandRunner`:
- No longer automatically add a 'help' option to commands that don't validate
  their arguments
- fix https://github.com/dart-lang/args/issues/123 (a slight variant of the PR https://github.com/dart-lang/args/pull/124)

The CommandRunner class currently doesn't work well with Commands that use an ArgParser.allowAnything() parser; this PR addresses that. It'll allow for subcommands that can do things like just delegate all arg parser (and execution) to another entity, like a subprocess.

### CODE REVIEW

File: test/test_utils.dart [hunk 1]
```diff
@@ -207,6 +207,22 @@ class AsyncCommand extends Command {
...
+  final name = 'allowAnything';
+
+  @override
+  final description = 'A command using allowAnything.';
+
+  @override
+  final argParser = ArgParser.allowAnything();
+
+  @override
+  Future run() => Future.value().then((_) => hasRun = true);
```
Comment (kevmoo) [inline test/test_utils.dart]: just use `async` here?

File: lib/command_runner.dart [hunk 2]
```diff
@@ -364,8 +364,10 @@ abstract class Command<T> {
   List<String> get aliases => const [];
 
   Command() {
-    argParser.addFlag('help',
-        abbr: 'h', negatable: false, help: 'Print this usage information.');
+    if (!argParser.allowsAnything) {
```
Comment (kevmoo) [inline lib/command_runner.dart]: Are we confident this won't break anyone?

File: test/test_utils.dart (hunk 1 above)
Comment (jakemac53) [reply test/test_utils.dart]: you would need to `await null` before this line if you do that.

File: lib/command_runner.dart (hunk 2 above)
Comment (jakemac53) [reply lib/command_runner.dart]: If I recall the `allowAnything` arg parser never worked with a `Command` anyways. So I think this is likely fine outside of maybe some weird edge cases where somebody implemented `ArgParser`?

**Comment by jakemac53** [comment]
We should ideally test this in flutter and internal before landing though as it is a bit risky

**Comment by kevmoo** [comment]
ack

**Comment by devoncarew** [comment]
Thanks for the review! Agree w/ the testing before a larger rollout.

Wrt flutter, I think this will allow them to improve the way they expose things like pub. Currently w/ the way their command dispatch is implemented, if you run something like `flutter pub upload -h` you don't get any of the `uploader` options printed even thought uploader has ~2 options. Changing to this style delegating would fix that. (cc @jonahwilliams for fyi)

File: test/test_utils.dart (hunk 1 above)
Comment (natebosch) [reply test/test_utils.dart]: Why are we relying on there being a microtask before setting the value?

Comment (devoncarew) [reply test/test_utils.dart]: This is taken from the command impl. above, but it looks like it may not be necessary.

File: lib/command_runner.dart (hunk 2 above)
Comment (devoncarew) [reply lib/command_runner.dart]: If they were using the `allowAnything` parser, the `argParser.addFlag()` call below would throw an exception (so, I don't think any existing code will be broken here).

File: test/test_utils.dart (hunk 1 above)
Comment (devoncarew) [reply test/test_utils.dart]: (changed to just `void run() { hasRun = true; }`)

**Comment by devoncarew** [comment]
Landing this; will likely publish as a dev release, and test that through the sdk/flutter/google3 repos.

**Comment by natebosch** [comment]
> will likely publish as a dev release, and test that through the sdk/flutter/google3 repos.

The Dart SDK and google3 don't need a version published in pub.
